### ISSUE TITLE
Uncaught exception when using this gem on iOS

### ISSUE THREAD

**Issue by WaveformDelta** [take-cheeze/mruby-marshal#20]
I am using this gem with the [ios-embedded-ruby](https://github.com/carsonmcdonald/ios-ruby-embedded) project and including it in the [MRubyiOSExample](https://github.com/carsonmcdonald/MRubyiOSExample) app. However, including this (or any C++ gem) into the project causes a problem with exception handling when I run it on the device. I can cause an unrecoverable crash with code as simple as:

```
begin
  raise StandardError, "This is an example error"
rescue => ex
  p "Caught error: #{ex.class}, #{ex.message}"
end
```

On a device, the ``raise`` statement will crash with a SIGABRT on the call to ``MRB_THROW`` in the ``mrb_exc_raise()`` method in the ``error.c`` source file with the message:

```
libc++abi.dylib: terminating with uncaught exception of type int
```

**Additional information**
- This only happens on a device, not a simulator (note that ``ios-embedded-ruby`` builds for both targets)
- This is not specific to this gem, but seems to be a problem with building C++ code for iOS; I verified this by making a 'dummy' gem that uses C++ and including only that, and got the same behavior; if I build it as C, the crash doesn't happen
- To get this ``mruby-marshal`` gem to build for iOS, I had to replace the ``mruby-onig-regexp`` gem dependency with the ``mruby-pure-regexp`` gem
- I added the 'conf.cxx' flags to the ``ios-ruby-embedded`` Rakefile (copy of the 'conf.cc' flags)
- I tried the ``enable_cxx_exception`` and ``enable_cxx_abi`` options in the Rakefile, but did not find a combination that solved the problem
- I had to add the ``-lc++`` flag to "Other Linker Flags" in the Xcode project
- I have posted this issue to the [ios-ruby-embedded](https://github.com/carsonmcdonald/ios-ruby-embedded/issues/12) project and [StackOverflow](https://stackoverflow.com/questions/45892243/mrb-throw-causes-error-in-libcabi-dylib) as well

**Comment by take-cheeze** [comment]
I don't know what version of `mruby` you are using and full `Rakefile` that cause this issue,
so I can't say much about it.
Customization isn't bad though makes things hard to verify so I need more information.

As I see, `conf.cxx` and `conf.linker` should be `clang++` or `-x c++`(force compiler to treat source as C++ code) should be added.
Add `-fexceptions` to your **C** compiler flags.
I prefer `enable_cxx_exception` since most C project aren't for C++ compilers.

**Comment by WaveformDelta** [comment]
Adding ``-fexceptions`` was the piece I was missing: with that added, all code now works as expected.

I created a [Gist](https://gist.github.com/WaveformDelta/a41d70d3a715c6735806b4d7f2c07fb2) with my modified ``Rakefile``; I included the ``-x c++`` argument to clang even though I found in testing I don't need it.  For now, I'm not using the ``enable_cxx_exception``.

I accept this issue as closed. Thank you for your help!!
